### ISSUE TITLE
.NET Core Preview 7 - Child node "2" exited prematurely / StackOverflow Exception on Build 

### ISSUE THREAD

**Issue by peter-bull** [microsoft/msbuild#4560]
# Issue Title

Child node "2" exited prematurely / StackOverflow Exception on Build with .NET Core 3.0 Preview 7 when referencing a WPF Class Library Project from an ASP.NET Core 3.0 Project

# General

I have a project which has a shared Class Library Project that uses the Windows Desktop SDK, this library which is used by a WPF Desktop Project (to offer desktop based ui but with functionality in core class library) and a ASP.NET Core Project (to offer web-based ui but re-used all other functionality) and this built in .NET Core 3.0 Preview 6 but does not build in .NET Core 3.0 Preview 7 - this can be easily replicated by creating a Class Library project and modifying the .csproj as follows:

```xml
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
	<UseWPF>true</UseWPF>
  </PropertyGroup>
</Project>
```

Then by creating a ASP.NET Core 3.0 Web Application and referencing this library, you don't need to create another project for WPF one as when referenced and built either a **Child node "2" exited prematurely** error will occur in **Visual Studio 2019 Preview 4** or a **StackOverflow** when doing **dotnet build** from the command line, this works in .NET Core Preview 6 without any errors but can be reproduced easily with just the two project types mentioned

Here's an extract from a file that was created during a build but have seen various errors but this one has occurred the most since upgrading Visual Studio 2019 Preview to 16.3.0 Preview 1.0

```
2019-07-29 08:50:36.654|00006892|
C:\build\aria-cpp-v1\clienttelemetry\src\LogManagerImpl.cpp(626): class Microsoft::Applications::Telemetry::ILogger *__thiscall Microsoft::Applications::Telemetry::LogManagerImpl::Initialize(const class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > &,const struct Microsoft::Applications::Telemetry::LogConfiguration &) WARNING: Invalid in-ram queue size (20971520), adjusted to max ram queue size
```

**Comment by LeeDumond** [comment]
I can confirm this as well. Worked fine in Preview 6.

Also, the output window says "Diagnostic information may be found in files in "C:\Users\Lee.Dumond\AppData\Local\Temp\" and will be named MSBuild_*.failure.txt" but this file is  NOT appearing at all.

**Comment by rainersigwald** [comment]
Thanks for the report! I followed your instructions to create a repro zip: 
[msbuild4560.zip](https://github.com/microsoft/msbuild/files/3443020/msbuild4560.zip)


```sh-session
$ dotnet build
Microsoft (R) Build Engine version 16.3.0-preview-19369-03+db6ae5a0d for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 84.2 ms for S:\repro\microsoft\msbuild\issues\4560\WebApp\WebApp.csproj.
  You are using a preview version of .NET Core. See: https://aka.ms/dotnet-core-preview
  LibWithWpf -> S:\repro\microsoft\msbuild\issues\4560\LibWithWpf\bin\Debug\netcoreapp3.0\LibWithWpf.dll
Stack overflow.
```

There are several problems here:

- [ ] The stack overflow that's causing the problem: aspnet/AspNetCore#12693
- [ ] "Stack overflow" has no diagnostic information whatsoever.
- [ ] Whatever's causing the worker node crash (not reproduced via `dotnet build WebApp.csproj` in the repro, but reproduced via `dotnet build 4560.sln`)

I think the latter two are related and caused by poor recovery from a `StackOverflowException`: the process just dies with `COR_E_STACKOVERFLOW` (`0x800703e9`), so we can't catch the exception and handle it to report a graceful error.

Under a debugger it's easier to see what's going on: infinite recursion in `Microsoft.NET.Sdk.Razor.Tasks.dll!Microsoft.AspNetCore.Razor.Tasks.ReferenceResolver.Resolve(Microsoft.AspNetCore.Razor.Tasks.ReferenceResolver.ClassifiedAssemblyItem classifiedItem)`. I created aspnet/AspNetCore#12693 to track that with the owners of the relevant task.

I'm going to leave this open to see if we can do anything in MSBuild to make this more obvious, but I'm not optimistic at the moment since we were killed by the CLR when encountering the stack overflow.
